pandas/core/arrays/base.py

```diff
@@ -610,3 +615,91 @@ def _ndarray_values(self):
         used for interacting with our indexers.
         """
         return np.array(self)
+
+
+def ExtensionOpsMixin(include_arith_ops, include_logic_ops):
+    """A mixin factory for creating default arithmetic and logical operators,
+    which are based on the underlying dtype backing the ExtensionArray
+
+    Parameters
+    ----------
+    include_arith_ops : boolean indicating whether arithmetic ops should be
+                        created
+    include_logic_ops : boolean indicating whether logical ops should be
+                        created
+
+    Returns
+    -------
+    A mixin class that has the associated operators defined.
+
+    Usage
+    ------
+    If you have defined a subclass MyClass(ExtensionArray), then
+    use MyClass(ExtensionArray, ExtensionOpsMixin(True, True)) to
+    get both the arithmetic and logical operators
+    """
+    class _ExtensionOpsMixin(object):
+        pass
+
+    def create_method(op_name):
+        def _binop(self, other):
+            def convert_values(parm):
+                if isinstance(parm, ExtensionArray):
+                    ovalues = list(parm)
+                elif is_extension_array_dtype(parm):
+                    ovalues = parm.values
+                elif is_list_like(parm):
+                    ovalues = parm
+                else:  # Assume its an object
+                    ovalues = [parm] * len(self)
+                return ovalues
+            lvalues = convert_values(self)
+            rvalues = convert_values(other)
+
+            # Get the method for each object.
+            def callfunc(a, b):
+                f = getattr(a, op_name, None)
+                if f is not None:
+                    return f(b)
+                else:
+                    return NotImplemented
+            res = [callfunc(a, b) for (a, b) in zip(lvalues, rvalues)]
```

Comment:
Orthogonal to @jorisvandenbossche 's comment, instead of `[callfunc(a, b) for (a, b) in zip(lvalues, rvalues)]` consider `libops.vec_binop(lvalues, rvalues, callfunc)`.  Should be equivalent but more performant.

Reply:
That doesn't seem to work properly?

```
In [64]: a = np.random.randn(100000).astype(object)

In [65]: b = np.random.randn(100000).astype(object)

In [66]: def func(a, b):
    ...:     return a + b
    ...: 
    ...: 

In [67]: a + b
Out[67]: 
array([-1.7373671804522872, -0.44608341233368065, -3.5963900176223706, ...,
       -1.0538754426986237, -2.0261407625013654, -0.6206810372160029], dtype=object)

In [68]: pd._libs.ops.vec_binop(a, b, func)
Out[68]: array([nan, nan, nan, ..., nan, nan, nan], dtype=object)

In [69]: np.array([func(aa, bb) for aa, bb in zip(a, b)], dtype=object)
Out[69]: 
array([-1.7373671804522872, -0.44608341233368065, -3.5963900176223706, ...,
       -1.0538754426986237, -2.0261407625013654, -0.6206810372160029], dtype=object)
```

Reply:
Ah, that's because it does a `maybe_convert_bool` at then end, I suppose. Is there a version that does not do this?

Reply:
Ahh, I guess that explains the "bin" part of "vec_binop".  So it looks like a) the original suggestion is incorrect and b) it would probably be easy to implement and may improve perf.

Reply:
I thought "binary" meant that it is an op between *2* values, not that the result in boolean. So I think the naming is just confusing.

Reply:
Agreed on both counts.